base/essentials.jl

```diff
@@ -783,3 +766,108 @@ ismissing(::Missing) = true
 
 function popfirst! end
 function peek end
+
+"""
+    @__LINE__ -> Int
+
+`@__LINE__` expands to the line number of the location of the macrocall.
+Returns `0` if the line number could not be determined.
+"""
+macro __LINE__()
+    return __source__.line
+end
+
+# Just for bootstrapping purposes below
+macro __FILE_SYMBOL__()
+    return Expr(:quote, __source__.file)
+end
+
+# Iteration
+"""
+    isdone(itr, state...) -> Union{Bool, Missing}
+
+This function provides a fast-path hint for iterator completion.
+This is useful for mutable iterators that want to avoid having elements
+consumed, if they are not going to be exposed to the user (e.g. to check
+for done-ness in `isempty` or `zip`). Mutable iterators that want to
+opt into this feature shoud define an isdone method that returns
+true/false depending on whether the iterator is done or not. Stateless
+iterators need not implement this function. If the result is `missing`,
+callers may go ahead and compute `iterate(x, state...) === nothing` to
+compute a definite answer.
+"""
+isdone(itr, state...) = missing
+
+"""
+    iterate(iter [, state]) -> Union{Nothing, Tuple{Any, Any}}
+
+Advance the iterator to obtain the next element. If no elements
+remain, nothing should be returned. Otherwise, a 2-tuple of the
+next element and the new iteration state should be returned.
+"""
+function iterate end
+
+# Compatibility with old iteration protocol
+function iterate(x, state)
+    @_inline_meta
+    done(x, state) && return nothing
+    return next(x, state)
+end
+const old_iterate_line_prev = (@__LINE__)
+iterate(x) = (@_inline_meta; iterate(x, start(x)))
+
+struct LegacyIterationCompat{I,T,S}
+    done::Bool
+    nextval::T
+    state::S
+    LegacyIterationCompat{I,T,S}() where {I,T,S} = new{I,T,S}(true)
+    LegacyIterationCompat{I,T,S}(nextval::T, state::S) where {I,T,S} = new{I,T,S}(false, nextval, state)
+end
+
+function has_non_default_iterate(T)
+    world = ccall(:jl_get_world_counter, UInt, ())
+    mt = Base._methods(iterate, Tuple{T}, -1, world)
+    # Check if this is the above method
+    if (mt[1][3].file == @__FILE_SYMBOL__) && (mt[1][3].line == old_iterate_line_prev + 1)
+        return false
+    end
+    return true
+end
+
+const compat_start_line_prev = (@__LINE__)
+function start(itr::T) where {T}
+    has_non_default_iterate(T) || throw(MethodError(iterate, (itr,)))
+    y = iterate(itr)
+    y === nothing && return LegacyIterationCompat{T, Union{}, Union{}}()
+    val, state = y
+    LegacyIterationCompat{T, typeof(val), typeof(state)}(val, state)
+end
+
+function next(itr::I, state::LegacyIterationCompat{I,T,S}) where {I,T,S}
```

Comment:
This (and similar for `done`) are real ambiguity magnets for all iterators out there that define `next(it::MyIterator, state)`. Of course, `next` should never be called with arguments that hit the ambiguity, but `detect_ambiguities` is not amused. But I suppose there is no easy way to prevent that?